src/main/scala/sigmastate/lang/SigmaBuilder.scala

```diff
@@ -0,0 +1,171 @@
+package sigmastate.lang
+
+import sigmastate.Values.Value
+import sigmastate._
+import sigmastate.lang.Constraints.{TypeConstraint2, onlyNumeric2, sameType2}
+import sigmastate.lang.Terms._
+import sigmastate.lang.exceptions.{BuilderException, ConstraintFailed}
+import sigmastate.serialization.OpCodes
+
+trait SigmaBuilder {
+
+  def EQ[S <: SType](left: Value[S], right: Value[S]): EQ[S]
```

Comment:
@greenhat we need to further abstract from internal workings of builders by hiding actual return type.
So I suggest to have signatures like this `def EQ[S <: SType](left: Value[S], right: Value[S]): Value[SBoolean.type]`
We can add synonims `type SBoolValue = Value[SBoolean.type]` to make life better.
Also why not to use typical naming for type arguments like `T,U,V,R` or `A,B,C`.
However `S` is also ok.